### ISSUE TITLE
[Question] How to handle build of multiple projects?

### ISSUE THREAD

**Issue by second-reality** [mesonbuild/meson#6404]
### Dear meson developers,

I successfully ported build system of my company from Make to Meson (~50 binaries in different projects, custom targets, generators, multiple languages). I have been very pleased by Meson. I added all our code (from multiple git submodules) in a single repo, and implemented a monolithic (and massive) meson.build file. So far, so good.

My coworkers adopted Meson, and for a new project, we would like to use it as the native build system (instead of make). However, some people prefer to clone only one project, instead of having whole repository (with submodules). 

**I want to allow to build a project, either as a separate project, or as part of my monolithic build. And I didn't find a correct solution to do it.**

-------------------------------------------------------------------------------------

### Concrete example

Imagine I have a project A and B. I want both of them to be buildable independently or together.

$ tree
```
├── A
│   ├── meson.build
│   └── src
│       ├── a.cpp
│       └── meson.build
├── B
│   ├── meson.build
│   └── src
│       ├── b.cpp
│       └── meson.build
└── meson.build

```
$ cat A/meson.build
```
project('a', 'cpp')
subdir('src')

```
$ cat A/src/meson.build
`executable('a', ['a.cpp'])`

-- here is my toplevel meson build file
$ cat meson.build
```
project('all', 'cpp')
subdir('A/src')
subdir('B/src')
```

By doing this, I can easily build A, B, or both of them.

Alas, I need both binaries a and b to be directly under my build folder. With this solution, I have this hierarchy:
$ tree build
```
build
├── A
│   └── src
│       ├── 4f2b75d@@a@exe
│       │   └── a.cpp.o
│       └── a
├── B
│   └── src
│       ├── 862d3b9@@b@exe
│       │   └── b.cpp.o
│       └── b
├── build.ninja
```

Main problem is that, to build only A or B, you need to fully specify path to it. Which is not very convenient.

$ ninja -C build a
```
ninja: Entering directory `build'
ninja: error: unknown target 'a', did you mean 'all'?
```
_Expected result would be to be able to build 'a' like this, and that it would be available as 'build/a'_

$ ninja -C build A/src/a
```
ninja: Entering directory `build'
ninja: no work to do. 
```

---------------------------------------------------------------------------------------------------
### Other ideas

I tried to hack around meson.build_root() + 'exe_name', but I still have the issue with ninja (and a warning saying that what I do is wrong!). So, I shamely copy/pasted build of both project in my monolithic meson.build. This is boring, especially because paths are not the same (hierarchy has one level more), and when you change somewhere, you have to think at the other place.

I could as well define only list of sources in meson.build src directories, and copy/paste only the executables definition, but it still feels wrong to me.

I searched extensively in documentation, and a little bit in source code, without finding what I wanted.

So, what is the way to do it? Is there a way to define a kind of "alias" for my subdirs build targets? Or else, how am I supposed to organize build system to do this?

I don't want to use subproject, because it does not match my need there. All my projects are more or less independent.

**Thank you for your answers and this delightful build system!**

**Comment by xclaesse** [comment]
This is called subprojects: https://mesonbuild.com/Subprojects.html

In short, you create a main project that contains `subprojects/A/meson.build`, `subprojects/B/meson.build`, etc. The main project calls `subproject('A')` and `subproject('B')`.

**Comment by second-reality** [comment]
Hello Xavier, as I said in the end of my message, subproject is not what I need.

It results in the same problem I have, with hierarchy being represented build directory, when I would like to have a "flat" hierarchy. (build/a, and 'a' target available in ninja).

build/
├── build.ninja
├── compile_commands.json
└── subprojects
    └── A
        ├── a
        └── b400d09@@a@exe
            └── a.cpp.o

**Comment by mensinda** [comment]
I would argue that this is *exactly* what subprojects are for. See [gst-build](https://gitlab.freedesktop.org/gstreamer/gst-build) for a project that does something very similar.

If you want all the outputs in the same directory, you could have a look at the [`layout`](https://mesonbuild.com/Builtin-options.html#core-options) meson option. If you use `meson -Dlayout=flat ...`, meson will put all outputs into the same directory (`meson-out`). This will also make the target `a` (from your example) available in ninja (although all outputs have a `meson-out` prefix).

Using `subdir` for this will not work, because there has to be **exactly one** `project` call at the beginning of each meson tree (`subdir` basically `#include`s the other `meson.build` files, thus no `project` call is allowed in a `subdir`). The only exception here are subprojects since they are treated differently and are more sandboxed. So there are only the following options for your use case:

|                                     | Monolithic repo    | Independently buildable |
| ----------------------------------- | :----------------: | :---------------------: |
| `project` in each component         | :x:                | :white_check_mark:      |
| `project` in monolith with `subdir` | :white_check_mark: | :x:                     |
| Subprojects                         | :white_check_mark: | :white_check_mark:      |

**Comment by xclaesse** [comment]
I still believe it is subprojects you need to use here. You can give a try to flat layout as mentioned above. If ninja target names is important, you can use [alias_target()](https://mesonbuild.com/Reference-manual.html#alias_target).

I guess one issue you could have with subprojects is you must move their source code into `subprojects/` subdirectory, instead of having them at the root as you currently do. I personally think that's a limitation we should remove for subprojects that have the code inside the main project repository (not using .wrap files). But I know that idea has opposition inside the Meson developers.

**Comment by second-reality** [comment]
Hello Daniel and Xavier,

I didn't see the layout option, but it seems like it answers exactly my need!
Concerning subprojects, you convinced me to adopt it! Restriction on having a "subprojects" directory seems a bit rigid to me. But I solved the issue by leaving the code where it is, and just symlink projects in subprojects directory to make meson happy.

However, I still have the problem concerning targets name in ninja. After adopting flat layout, I have those targets:

```
user@pc:/home/user/try/build> ninja -t targets all
...
meson-out/a: cpp_LINKER
...
meson-out/b: cpp_LINKER
...
```

So I miss a and b targets. Having a fix meson-out directory seems a bit rigid to me.

I finally solved the issue with alias_target (exactly what I was looking for!), to the price of adding variables for each binary.

```
$ cat A/meson.build
project('a', 'cpp')
a = executable('a', 'a.cpp')

$ cat meson.build
project('all', 'cpp')
A = subproject('A')
B = subproject('B')

alias_target('a', A.get_variable('a'))
alias_target('b', B.get_variable('b'))
```

So, I finally get what I want! (with a directory meson-out in more to access binaries, but that is fine)

Do you think a patch to make flat layout to not introduce another directory (after all, build directory is already there for this, no?) would be accepted? I can work around this if needed.
Maybe via a new layout (flat_really_flat :)?)

Thank you for your time.

**Comment by second-reality** [comment]
I tried this morning solution said above.

Alas, having meson-out folder makes that configured file (with configure) and custom targets are not built in the good directory! In more, some dependencies are missing.

So, I think it would be very nice to have a **really** flat layout.

Do you think a proposal around this would be accepted?
Here is a commit doing this I hacked this morning:
https://github.com/second-reality/meson/commit/0e3020c25af5c657d05341d0c04e404dd23fa39a

**Comment by QuLogic** [comment]
Why do you need them in the same directory? I think you are trying to solve the wrong problem.

**Comment by second-reality** [comment]
Hi Elliott,

In our codebase, we have binaries built with executable(), and binaires built with custom target as well, basically this kind of things: https://aomedia.googlesource.com/aom/

Since we don't want to convert this build system in meson (we just use the binary in some scripts, we don't develop inside it!), we build it via a custom_target. So far so good.

With flat=layout option, I obtain this:
```
$ tree
.
├── aomdec-generic.exe
├── aomenc-generic.exe
├── build.ninja
├── compile_commands.json
├── meson-out
│   ├── my_binary
│   ├── my_other_binary
```

So in the end, binaries are built in several directories. How does the final user is suppose to guess where to find his binary?

And since custom_target do not accept subdirectories in their names, it is *impossible* to have everything in the same dir in the end.

IMHO, flat layout + subprojets answers the need to build several projects independently, or together. However, the introduction of meson-out directory breaks what would be logic. I don't expect someone that use meson to guess binaries are in build/meson-out instead of simply build folder.

About your question, why do we need them in the same directory? We have some tools that know how our sandbox is organized, and especially where to find built binaries.

Basically, like this:

```
get_build_dir() {
  [ $# -ne 1 ] && die "usage: build_id"

  build_id=$1
  shift

  echo "$(get_sandbox_root)/$build_directory_name/$build_id"
}
```

So, I'm fine with adding meson-out to constructed path, but definitely not fine to search in two directories at the same time.

Either everything should be in meson-out, or nothing. Plus, meson-out creates a sublevel in targets for ninja. And yes, even if ninja is fast to build everything, there is a use case to only build a partial sandbox when you have many binaries (Continuous integration of some projects for instance).

Does that sounds right to you?

**Comment by QuLogic** [comment]
> So in the end, binaries are built in several directories. How does the final user is suppose to guess where to find his binary?

The final user installs the project; they shouldn't worry about how it looks in the build directory. The install prefix is consistent, the build directory is irrelevant.

**Comment by second-reality** [comment]
That is a good point you got there.

We don't install our binaries, and work straight from build folder. Maybe that is the root of our problems.

Why we don't do this? When you develop, it can be very (very very) error prone to have two different folders where your binaries are (forgot install, oops!).
In more, it requires more disk space, which is not always convenient when you have 50+ binaries in debug mode.

**Comment by xclaesse** [comment]
When running uninstalled, you often need to set some env variables. In GStreamer project we run without installing, but you have to run your shell with a wrapper script: https://gitlab.freedesktop.org/gstreamer/gst-build/blob/master/gst-env.py. It sets PATH to the location of the various executables in build dir, etc.

I started working on integrating that directly into Meson: https://github.com/mesonbuild/meson/pull/6229

**Comment by taozuhong** [comment]
Share my solution, it cost me 1.5 days:
```
├── A
│   ├── meson.build
│   └── src
│       ├── a.cpp
│       └── meson.build
├── B
│   ├── meson.build
│   └── src
│       ├── b.cpp
│       └── meson.build
├── C
│   ├── meson.build
│   └── src
│       ├── b.cpp
│       └── meson.build
└── meson.build
```

Root meson.build
```
project('all', 'c')
subdir('A')  # library A, layout: mirror
subdir('B')  # library B, layout: mirror
subdir('C')  # execute C, layout: mirror
```

Some of C meson.build
```
# Copy library for debug mode
lib_extension = ''
if host_os.contains('mingw') or host_os.contains('cygwin') or host_os.contains('windows')
  lib_extension = '.dll'
elif host_os.contains('linux')
  lib_extension = '.so'
elif host_os.contains('darwin') or host_os.contains('machten') or host_os.contains('rhapsody')
  lib_extension = '.dylib'
endif

custom_target('libA',
  command: ['cp', '@BUILD_ROOT@/A/libA' + lib_extension, '@BUILD_ROOT@/C'],
  input: libA,
  output: 'libA' + lib_extension,
  install: false,
  build_by_default: true,
  depends: [libA, appC]
)

custom_target('libB',
  command: ['cp', '@BUILD_ROOT@/B/libB' + lib_extension, '@BUILD_ROOT@/C'],
  input: libB,
  output: 'libB' + lib_extension,
  install: false,
  build_by_default: true,
  depends: [libB, appC]
)
```

**Comment by second-reality** [comment]
Thanks for sharing it.

Since our build as 50+ targets, defining aliases or custom targets for all of them is not a solution we want to follow.

We decided to migrate to a monorepo, and we switch to a single meson project with all targets instead of trying to support build per subproject.

**Comment by taozuhong** [comment]
I have to split the project into multi libraries because the project has 400+ source files, a minor change will be raised to rebuild the entire project, the building costs a lot of time.

**Comment by marc-h38** [comment]
> We decided to migrate to a monorepo,

You may find this interesting https://www.bitquabit.com/post/unorthodocs-abandon-your-dvcs-and-return-to-sanity/

**Comment by second-reality** [comment]
Nice reading.

In our case, this migration really solved our dev issues, but our codebase is "small" and we don't reach any critical size (regarding git/meson).

In more, a monorepo really takes big effort regarding CI and tooling in general. For me, it's definitely worth, but someone has to make it's hand dirty.

I didn't investigate time yet in bazel/buck or any "next-gen" build system (I love meson simplicity and it does the job), but having a true cache handling could be an amazing addition to meson (or ninja more exactly). Ccache helps on this, but it only handles C/C++ (plus, it's hard to make it shared among devs).

**Comment by eli-schwartz** [comment]
The current release candidate of meson 0.61.0 adds support for autodetecting sccache the same way ccache is currently supported.

This adds rust and cuda on top of C/C++, and also lets you distribute the cache using for example S3.

**Comment by second-reality** [comment]
Very nice!

We use D a lot (70% of our codebase), and alas, there is no equivalent.

That's a place where a language agnostic cache system could be nice. By cache, I mean ccache functionality but server support as well (like bazel does).

**Comment by QuLogic** [comment]
> We don't install our binaries, and work straight from build folder. Maybe that is the root of our problems.

The [new devenv support](https://mesonbuild.com/Release-notes-for-0-58-0.html#developer-environment) may solve this?

**Comment by second-reality** [comment]
Sounds good. We already have a thing like this (via a script shell, that does build + set path + run), but it's good to see it's integrated in meson.

I'll close the issue, since it was originally oriented on use of subprojects, which is something we dropped completely.
